File: backend/src/main/java/io/ntf/api/activity/model/ActivityRepository.java
package io.ntf.api.activity.model;

import org.springframework.data.mongodb.repository.ReactiveMongoRepository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.time.LocalDateTime;

public interface ActivityRepository extends ReactiveMongoRepository<Activity, String> {

  Flux<Activity> findByUserIdOrderByStartDesc(String userId);

  Flux<Activity> findByUserIdAndStartIsAfterOrderByStartDesc(String userId, LocalDateTime dateTime);

  Mono<Activity> findByUserIdAndId(String userId, String id);

  Mono<Activity> findByUserIdAndEndNull(String userId);
}

File: backend/src/main/kotlin/io/ntf/api/activity/model/ActivityRepository.kt
package io.ntf.api.activity.model

import org.springframework.data.mongodb.repository.ReactiveMongoRepository
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

import java.time.LocalDateTime

interface ActivityRepository : ReactiveMongoRepository<Activity, String> {

  fun findByUserIdOrderByStartDesc(userId: String): Flux<Activity>

  fun findByUserIdAndId(userId: String, id: String): Mono<Activity>

  fun findByUserIdAndEndNull(userId: String): Mono<Activity>

  fun findByUserIdAndStartBetweenOrderByStartDesc(userId: String, from: LocalDateTime, to: LocalDateTime): Flux<Activity>

  fun findByUserIdAndTags(userId: String, tags: List<String>): Flux<Activity>
}
